app/services/ProcessCsvService.scala

```diff
@@ -0,0 +1,273 @@
+import akka.util.ByteString
+import config.ApplicationConfig
+import connectors.ERSFileValidatorConnector
+import models.upscan.UpscanCsvFileData
+import models.{CsvFileContents, ERSFileProcessingException, SchemeData, SchemeInfo, SubmissionsSchemeData, CsvFileSubmissions}
+import play.api.Logger
+import play.api.mvc.Request
+import services.ERSTemplatesInfo.ersSheets
+import services.FlowOps.eitherFromFunction
+import services.audit.AuditEvents
+import services.validation.ErsValidator.getCells
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.services.validation.DataValidator
+import uk.gov.hmrc.services.validation.models._
+import utils.{ErrorResponseMessages, ValidationUtils}
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+import javax.inject.{Inject, Singleton}
+
+@Singleton
+class ProcessCsvService @Inject()(auditEvents: AuditEvents,
+                                  dataGenerator: DataGenerator,
+                                  appConfig: ApplicationConfig,
+                                  ersConnector: ERSFileValidatorConnector
+                                 )(implicit executionContext: ExecutionContext,
+                                   actorSystem: ActorSystem) {
+
+  private val uploadCsvSizeLimit: Int = appConfig.uploadCsvSizeLimit
+  private[services] val ersSheetsClone: Map[String, SheetInfo] = ersSheets
+
+  def extractEntityData(response: HttpResponse): Source[ByteString, _] =
+    response match {
+      case HttpResponse(akka.http.scaladsl.model.StatusCodes.OK, _, entity, _) => entity.withSizeLimit(uploadCsvSizeLimit).dataBytes
+      case notOkResponse =>
+        Logger.error(
+          s"[ProcessCsvService][extractEntityData] Illegal response from Upscan: ${notOkResponse.status.intValue}, " +
+            s"body: ${notOkResponse.entity.dataBytes}")
+        Source.failed(
+          ERSFileProcessingException(
+            s"${ErrorResponseMessages.fileProcessingServiceFailedStream}",
+            s"${ErrorResponseMessages.fileProcessingServiceBulkEntity}"))
+    }
+
+  def extractBodyOfRequest: Source[HttpResponse, _] => Source[Either[Throwable, List[ByteString]], _] =
+    _.flatMapConcat(extractEntityData)
+      .via(CsvParsing.lineScanner())
+      .via(Flow.fromFunction(Right(_)))
+      .recover {
+        case e => Left(e)
+      }
+
+  def processFiles(callback: UpscanCsvFileData, source: String => Source[HttpResponse, _])(
+    implicit request: Request[_], hc: HeaderCarrier
+  ): List[Future[Either[Throwable, CsvFileContents]]] =
+    callback.callbackData map { successUpload =>
+
+      val sheetName = successUpload.name.replace(".csv", "")
+      val tryValidator: Either[Throwable, (DataValidator, SheetInfo)] = dataGenerator.getValidatorAndSheetInfo(sheetName, callback.schemeInfo)
+
+      tryValidator.fold(
+        throwable => Future.successful(Left(throwable)),
+        setValidator => {
```

Comment:
Marking this here. I think Dan made some comments on ers-checking-frontend and ers-submissions about the use of _.1 and _.2 so this could apply here.

Reply:
cheers! will tidy that up.